Scripts/Runtime/Entities/TaskDriver/TaskSet/TaskData/DataStream/DataSource/Data/AbstractData.cs

```diff
@@ -33,7 +34,17 @@ public static DataTargetID GenerateWorldUniqueID(IDataOwner dataOwner, Type abst
 
         private DataTargetID m_WorldUniqueID;
 
-
+        /// <summary>
+        /// Whether the underlying data has potentially been updated by something getting write access to it.
+        /// </summary>
+        public virtual bool IsDataInvalidated
+        {
+            //If the current Read dependency has changed from what we last stored, then someone has written here
+            get => m_AccessController
+                .GetDependencyFor(AccessType.SharedRead)
+                .Equals_NoBox(m_LastSharedReadAccessJobHandle);
+        }
+        
```

Comment:
This only works if nobody has called `AcquireAsync(AccessType.SharedRead)` since write access was gained.

I think we'd be better off having the caller pass in the read handle at the last time they requested read because data invalidation is relative to the perspective of the data consumer not the provider.

Reply:
I'm not sure I fully understand. I've written out a flow below. Am I missing something? 

I think we are handling this where `AbstractData` is the consumer and handles storing the read handle for the last time it requested read. 


TDA tries to read from AbstractData
- LastSharedReadAccess set to the Data's Read Dependency (1)

TDB tries to read from same AbstractData
- LastSharedReadAccess set to the Data's Read Dependency (still 1)

TDC tries to write to the same AbstractData
- Data's dependencies move up the read is now (2)

TDD tries to read from the same AbstractData
- Data is invalid because the AbstractData's LastSharedReadAccess that was stored is not the same anymore.
- LastSharedReadAccess set to the Data's Read Dependency (now 2)

Reply:
After your last point what if TDA then tries to read from AbstractData?

It doesn't know that the data has changed since its last read and will think that `IsDataInvalidated`

From what I'm understanding, this results in one member's read operation hiding the data change from all other potential readers. It's only the first member that reads after a write that gets to know that the data is invalidated.

Reply:
Ugh... you're right!

Good point. I've moved the storage and checking of the read handles to where the caller for the jobs are which makes sense. We really want to know if we've been written to since the last time we ran this specific job. If we have, then our job needs to run again. If we haven't, then nothing has changed and so we don't need to run our job.